codepropertygraph/src/main/scala/io/shiftleft/passes/ParallelCpgPass.scala

```diff
@@ -109,12 +111,154 @@ abstract class ParallelCpgPass[T](cpg: Cpg, outName: String = "", keyPools: Opti
+
+  override def createAndApply(): Unit = createApplySerializeAndStore(null)
+
+  override def createApplySerializeAndStore(serializedCpg: SerializedCpg,
+                                            inverse: Boolean = false,
+                                            prefix: String = ""): Unit = {
+    import ConcurrentWriterCpgPass.producerQueueCapacity
+    baseLogger.info(s"Start of enhancement: $name")
+    val nanosStart = System.nanoTime()
+    var nParts = 0
+    var nDiff = 0
+
+    init()
+    val parts = generateParts()
+    nParts = parts.size
+    val partIter = parts.iterator
+    val completionQueue = mutable.ArrayDeque[Future[DiffGraph]]()
+    val writer = new Writer(serializedCpg, prefix, inverse)
+    val writerThread = new Thread(writer)
+    writerThread.setName("Writer")
+    writerThread.start()
+    try {
+      try {
+        // The idea is that we have a ringbuffer completionQueue that contains the workunits that are currently in-flight.
+        // We add futures to the end of the ringbuffer, and take futures from the front.
+        // then we await the future from the front, and add it to the writer-queue.
+        // the end result is that we get deterministic output (esp. deterministic order of changes), while having up to one
+        // writer-thread and up to producerQueueCapacity many threads in-flight.
+        // as opposed to ParallelCpgPass, there is no race between diffgraph-generators to enqueue into the writer -- everything
+        // is nice and ordered. Downside is that a very slow part may gum up the works (i.e. the completionQueue fills up and threads go idle)
+        var done = false
+        while (!done) {
+          if (completionQueue.size < producerQueueCapacity && partIter.hasNext) {
+            val next = partIter.next()
+            //todo: Verify that we get FIFO scheduling; otherwise, do something about it.
+            //if this e.g. used LIFO with 4 cores and 18 size of ringbuffer, then 3 cores may idle while we block on the front item.
+            completionQueue.append(Future.apply {
+              val builder = DiffGraph.newBuilder
+              runOnPart(builder, next.asInstanceOf[T])
+              builder.build()
+            })
+          } else if (completionQueue.nonEmpty) {
+            val future = completionQueue.removeHead()
+            val res = Await.result(future, Duration.Inf)
+            nDiff += res.size
+            writer.queue.put(Some(res))
+          } else {
+            done = true
+          }
+        }
+        //
+      } finally {
+        writer.queue.put(None)
+        writerThread.join()
+        finish()
+      }
+    } finally {
+      // the nested finally is somewhat ugly -- but we promised to clean up with finish(), we want to include finish()
+      // in the reported timings, and we must have our final log message if finish() throws
+      val nanosStop = System.nanoTime()
+      MDC.put("time", s"${(nanosStop - nanosStart) * 1e-6}%.0f")
```

Comment:
Putting context values doesn't seem to make sense here since you do the same calculation inside the log message?

Reply:
Yeah. I didn't put the MDC here, I just ported it from the ParallelCpgPass because I did not know whether any monitoring scripts rely on it. But you're right, I'll remove it.

Reply:
removed it